designs/rerun_cli_commands.md

```diff
@@ -0,0 +1,85 @@
+When a command is rerun after an error, the CLI will skip the completed steps, restoring the state of the program.
+The checkpoint data would be stored in the `<clusterName>/generated` folder as a file named `<clusterName>-checkpoint.yaml`.
+
+Example checkpoint file:
+
+```yaml
+completedTasks:
+  bootstrap-cluster-init:
+    ExistingManagement: false
+    KubeconfigFile: test/generated/test.kind.kubeconfig
+    Name: test
+  ensure-etcd-capi-components-exist: null
+  install-capi: null
+  pause-controllers-reconcile: null
+  setup-and-validate: null
+  update-secrets: null
+  upgrade-core-components:
+    components: []
+  upgrade-needed: null
+```
+
+When the task finishes with no errors, we will add it as a completed task in the checkpoint file with the necessary information (if any) to allow the command to rerun successfully.
+
+```go
+if commandContext.OriginalError == nil {
+    // add task to completedTasks in checkpoint file
+}
+```
+
+When the user re-runs the command, we will check if the task exists as a completed task in the checkpoint file.
+If so, we will restore any necessary information from the checkpoint file and set the next task.
+
+```go
+for task != nil {
+    if _, ok := checkpointInfo.CompletedTasks[task.Name()]; ok {
+        // Restore any necessary information from the checkpoint file & set task to the next task
+        continue
+    }
+    logger.V(4).Info("Task start", "task_name", task.Name())
+    ...
+    ...
+}
+```
+
+### Other notable changes:
+* If the CLI fails to install CAPI objects to the bootstrap cluster, it deletes the bootstrap cluster.
+  This will obviously cause problems if we were to rerun the command at that point.
+  Therefore, in the `Restore()` method for the `CreateBootstrapClusterTask`, we will re-create the bootstrap cluster based on the cluster information in the `bootstrap-cluster-init` key from the checkpoint file if the bootstrap cluster does not exist.
+  An alternative to this would be to not delete the bootstrap cluster if `InstallCAPITask` fails & go straight to `CollectDiagnosticsTask`.
+
+
+* If the CLI fails when moving CAPI management, things start to get a bit tricky.
+  We could move CAPI management back to the source cluster if it fails, but this operation can be risky & potentially leave the cluster in an undesirable state.
+  Therefore, we will move this out of the scope of this initial MVP.
+  A deeper dive would be required to decide on a better way to roll back those changes if moving CAPI fails.
+  An idea for a future improvement is to backup the CAPI objects associated with the move using the `clusterctl backup` command. See [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/cmd/backup.go).
+  For now, if the cluster fails at this point (which does not happen terribly often), it will still fail, and we will handle that case in the future.
+
+
+* The only task that we should always run is the `setup-and-validate` task.
```

Comment:
We want to make sure that no other step sets up structs in a way that is used in the subsequent steps. I think that is what @chrisdoherty4 was referring to.